### PR TITLE
#966 Password reset changes

### LINKED ISSUE #966: Add Password reset
Will require 2 simple templates, and a small codebase change:

### 1st template:

- enter username* - check its a valid username in system (use isUnique endpoint)
- onSubmit -- verification link sent to email of username
- On verification:

~~### 2nd template~~ (went with 1-template solution as per discussion below and explanation in [PR](#975))
~~- enter new password field~~
~~- onSubmit -- change password~~

### Changes to codebase:
- add a "forgot password" link on login screen that links directly to first template application (above)
---

\* would be good to allow reset by email address, but we're not enforcing unique email addresses currently, so wouldn't be able to tell which account it is for.

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fixes #966 

Requires back-end PR https://github.com/openmsupply/application-manager-server/pull/563

Snapshot to test: [demo_password_reset.zip](https://github.com/openmsupply/application-manager-web-app/files/7135526/demo_password_reset.zip) (This is "demo_2021_09_07_1" with an added password reset template)

### Summary of code changes

- Added "Forgot my password" link to login screen
- Changed "UserRegister" file/component to "NonRegisteredLogin" to reflect additional functionality, in that it is called from more places now, with options as to where to re-direct after (nonRegistered) login
- If sessionId is present in URL query, this will always be passed to back-end in login and userInfo requests, and the same sessionId will be returned in the subsequent JWT. This is ensures that a sessionId can be added to a url which makes the user identifiable (even though they're nonRegistered at this point)

### The PasswordReset template

Mostly as discussed in #966, works as follows:

- Template has 2 stages
- Generated application serial is just a 24-char alphanumeric string, to further help with obscuring access
- In Stage 1, only "enter username" is visible. User enters their username and, if it is valid, they can submit the form
- On initial submit:
  - verification record created (with link back to the application form, with sessionId)
  - email sent to user with confirmation link
  - status => SUBMITTED
- On verification:
  - Increment Stage (to 2)
  - status => DRAFT
  - link in success notification back to page 1 of application
- User clicks link, they'll be logged in as nonRegistered (with existing sessionId), and now the form has an additional "Change password" input visible (and the username field is uneditable)
- On final submit
  - Password hash changed (modifyRecord)
  - Outcome => Approved

Notes:
- if we wish to use "email" rather than "username" (or both) to identify user, this will be a trivial change after making the email field unique and non-Null (but I suggest doing it later)
- I'll add the "Password reset" template to the templates repository once this PR is approved and merged

### CODE REVIEW

**Comment by nmadruga** [comment]
Starting to review this one now : ) Sorry for delay - too much going on

Just noticed a minor issue with validation - the field has the correct validation but the section has it as valid... I had entered some other invalid username before when I first got the validation error (correct) in both the page and section, and after changing to this the section got valid. Any idea on what could be the problem?

Screen Shot 2021-09-16 at 12 24 01 PM

**Comment by nmadruga** [comment]
Also - I forgot to change the user that I'm changing the password to have an email so getting that error reported on issue #962. Sigh - we need to fix this one.

**Comment by CarlosNZ** [comment]
> Also - I forgot to change the user that I'm changing the password to have an email so getting that error reported on issue #962. Sigh - we need to fix this one.

Yes -- I'll do that one next, should be pretty simple. Thanks for the reminder.

**Comment by CarlosNZ** [comment]
> Any idea on what could be the problem?

Yeah, the Password plugin has internal validation only, because we don't let the actual password ever go outside the plugin (otherwise it would get saved to the database). It should still inform the outer validation process of its status though, so not totally sure what's going on. I'll have a look -- if it's a quick fix to the password plugin, I should be able to add it in, but I suspect it might require a more substantial dig into the "revalidation" system (which I find quite confusing whenever I've looked at it).

**Comment by nmadruga** [comment]
Yes, fair enough.

**Review by nmadruga — APPROVED** [review]
The PR is great. Thanks for working on this one so quickly Carl! 

I have tested a few things and noted that all works accordingly:
- when session was different the verification failed
- after trying a second time the verification code gave me and error
- checked that going through the process of "forget my password" works fine to reset the user passowrd
- logging out after that and trying the other process (for nonRegistered users) to create account also works fine

> Snapshot to test: [demo_password_reset.zip](https://github.com/openmsupply/application-manager-web-app/files/7135526/demo_password_reset.zip) (This is "demo_2021_09_07_1" with an added password reset template)

That's very nice of you to have linked a template to test this. Thanks! I think it would be good if this PR could also add the same template to `basic_snapshot` so we have that already on a existing snapshot that we use to base other work

**Comment by CarlosNZ** [comment]
> Just noticed a minor issue with validation - the field has the correct validation but the section has it as valid... I had entered some other invalid username before when I first got the validation error (correct) in both the page and section, and after changing to this the section got valid. Any idea on what could be the problem?

Ohhh -- ignore my last comment, this isn't even a password field, it's just a text input with standard validation. 🙄

So, I think what's happening is the behaviour we've been having all along -- it's just that we haven't come across a page with a single element that has a validation expression on it.

The progress bar only gets full revalidation info when clicking next or Summary, but we don't normally notice because it's state is "In progress" rather than "complete" -- but I can confirm that when you have errors on a page, but with other fields not filled, you don't get progress bar errors until trying to move forward, which is consistent with what you're seeing:
Screen Shot 2021-09-16 at 3 02 38 PM

And I can cause the same error you're seeing with any application, like so:
Screen Shot 2021-09-16 at 3 11 30 PM

I also checked the ProgressBar state, and (i don't fully know why), but it's getting "Valid: true" until you try and click ahead (and cause a revalidate). I think there was some reason why it was designed like this, but it escapes me now.

I don't think this is a big problem, since it's a pretty edge case (and doesn't actually allow anything bad to happen), but feel free to make an issue investigate further if you think it's important.

Good find, though!

File: src/contexts/UserState.tsx [hunk 1]
```diff
@@ -91,6 +92,9 @@ export function UserProvider({ children }: UserProviderProps) {
   const [state, dispatch] = useReducer(reducer, initialState)
   const userState = state
   const setUserState = dispatch
+  const {
+    query: { sessionId },
+  } = useRouter()
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/contexts/UserState.tsx]: Thanks. da281e48

**Comment by CarlosNZ** [comment]
> I think it would be good if this PR could also add the same template to `basic_snapshot` so we have that already on a existing snapshot that we use to base other work

Done in back-end (and copied to templates repo)

**Pull request merged**